src/utils/easings.hpp

```diff
@@ -0,0 +1,86 @@
+//
+// This is free and unencumbered software released into the public domain.
+//
+// Anyone is free to copy, modify, publish, use, compile, sell, or
+// distribute this software, either in source code form or as a compiled
+// binary, for any purpose, commercial or non-commercial, and by any
+// means.
+//
+// In jurisdictions that recognize copyright laws, the author or authors
+// of this software dedicate any and all copyright interest in the
+// software to the public domain. We make this dedication for the benefit
+// of the public at large and to the detriment of our heirs and
+// successors. We intend this dedication to be an overt act of
+// relinquishment in perpetuity of all present and future rights to this
+// software under copyright law.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+// OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+//
+//
+// For more information, please refer to <https://unlicense.org>
```

Comment:
While it's an OSI-approved license, the amount of hassle that an extra license is likely to cause the packagers seems greater than the amount of work it would take to reimplement the fairly trivial maths in this library, particularly if it was just reimplementing the functions that we'd actually use.

Starting with the packaging hassle, who are the authors?
Can we even comply with Debian's copyright tracking info, if we have a file where we don't know who the authors are?

Reply:
The code was copied from https://github.com/warrenm/AHEasing/blob/master/AHEasing/easing.c with modifications (namespace, replacing the C math macros, using `double` directly). I left the license alone, but I could just change it to our standard GPL. The original does say it's public domain.

Reply:
The original **doesn't** say it's public domain; the whole point about the Unlicense and CC0 is that there isn't a way to globally put stuff in the public domain.

Reply:
> The original doesn't say it's public domain

From the [site](https://unlicense.org) linked by the license:

> it is a template for dedicating your software to the [public domain](http://stpeter.im/writings/essays/publicdomain.html)

From the license itself:

> This is free and unencumbered software released into the public domain.

It is effectively public domain at least by some definition of it. It's best, of course, to either relicense the whole thing as you want (since the author is effectively disclaiming copyright, you can just do that without asking for permission) or, in order to abide by the author's original intentions, to keep the original license. The latter is what Vultraz did. What's the problem again?

Reply:
The main problem is that it's likely to cause work for every downstream who does manual or automated sanity-checks of the licensing, probably more work than just rewriting the code in this library.

Also, some jurisdictions don't give copyright holders a right to waive their copyright, which is why these sort of licenses exist. https://en.wikipedia.org/wiki/Public-domain-equivalent_license

Reply:
Steve, in all honesty, anything we do can create work for downstreams. That's the whole point of being an upstream. I feel you shouldn't focus too much on downstreams in this case but rather on whether there's any benefits for us in reusing existing PD code vs. writing our own version of it.

Regarding public domain in general, at least Debian [https://wiki.debian.org/DFSGLicenses#Public_Domain](does not mind it), even bearing in mind what you just said about its lack of universality.

I think it's fair to say that if the license's intention is to disclaim copyright then the work is effectively public domain and we can regard it as such for all intents and purposes until we get sued by the author. As for who the author(s) is(are), I suppose you can look at the commit history and check the [latest previous version under a different license than Unlicense](https://github.com/warrenm/AHEasing/blob/78083c37a69bd2e87307d5e1ed297b6d427ff99c/AHEasing/easing.c) to find out what the "copyright" statement says in that regard (note that it used to be licensed under the WTFPL).

Reply:
Note that this isn't actually the first time that public domain code has been added to Wesnoth's codebase: https://github.com/wesnoth/wesnoth/blob/1.16/src/libc_error.hpp

Reply:
Also I'm not sure what rewriting would look like - if it does what we need already, then "rewriting" would give you basically the same thing as this, but we also can't just remove the copyright information from these files either.

Reply:
I feel our README.md's "Licensing" section should give downstreams an accurate picture of how many licenses they'll find. We're currently saying that we do that work for them.

> Regarding public domain in general, at least Debian [does not mind it](https://wiki.debian.org/DFSGLicenses#Public_Domain), even bearing in mind what you just said about its lack of universality.

However, as that link says, they still want to be aware that the license is being used, and to have the author's name to list in the `debian/copyright` file; it would be good to include the name in the file instead of making the packager have to search the git commit logs or PR discussions for it.

I've now found that Debian have already discussed the Unlicense, and have accepted a package using it:
https://sources.debian.org/src/tvnamer/3.0.4-2/debian/copyright/
https://lists.debian.org/debian-legal/2021/12/msg00008.html

> I'm not sure what rewriting would look like - if it does what we need already, then "rewriting" would give you basically the same thing as this

That depends on how much of this is going to be used. If we're going to use the elastic and bounce functions then I see the reason for importing the library; however if we're only using linear, cubic and maybe another basic one (as I initially assumed) then writing functions as we use them seems less hassle than mentioning another license.

My assumptions about how much Vultraz plans to use have changed on that now that I've compared it to the source and seen how much work Vultraz did on adapting it to our coding style. I'd assumed it was a full-file cut&paste, rather than working though each individual function.